### ISSUE TITLE
when running JUnit5 tests from the maven cli specifying -Dtest= with a single test method, no tests are run

### ISSUE THREAD

**Issue by rotty3000** [bndtools/bnd#3580]
Here's an example from cli with trace enabled:
```
mvn verify -Pbnd-next -Dtest=org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1

[snip]

# framework=org.eclipse.osgi.launch.Equinox@399c4be1
# registered launcher with arguments for syncing
# will call main
# running in main thread
# test cases org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1
# run unresolved true
# using /home/rotty/projects/osgi-test/org.osgi.test.junit5/target/test-reports/test, path: /home/rotty/projects/osgi-test/org.osgi.test.junit5/target/test-reports/test
# received names of classes to test org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1
# testing request org.junit.platform.launcher.core.DefaultDiscoveryRequest@6a55299e
# TestPlan Execution Started: org.junit.platform.launcher.TestPlan@5553d0f5
# Execution Started: Bnd JUnit Platform Bundle Engine - [engine:bnd-bundle-engine]
# Execution Finished: Bnd JUnit Platform Bundle Engine - [engine:bnd-bundle-engine] - TestExecutionResult [status = SUCCESSFUL, throwable = null]
# TestPlan Execution Finished: org.junit.platform.launcher.TestPlan@5553d0f5

Test run finished after 9 ms
[         1 containers found      ]
[         0 containers skipped    ]
[         1 containers started    ]
[         0 containers aborted    ]
[         1 containers successful ]
[         0 containers failed     ]
[         0 tests found           ]
[         0 tests skipped         ]
[         0 tests started         ]
[         0 tests aborted         ]
[         0 tests successful      ]
[         0 tests failed          ]

[INFO] No Errors
```

**Comment by rotty3000** [comment]
See branch https://github.com/rotty3000/osgi-test-1/tree/junit5.initial

**Comment by rotty3000** [comment]
Do note that specifying just the class alone works as expected.

**Comment by kriegfrj** [comment]
Is there magic in m2e to turn the Maven "test=" into a "tester.names" OSGi property setting?

**Comment by kriegfrj** [comment]
> Is there magic in m2e to turn the Maven "test=" into a "tester.names" OSGi property setting?

Scratch that, I can see from the tester's trace output that the tester.names property has made it as far as the tester's Activator.

**Comment by kriegfrj** [comment]
Ok, I tried your setup using your branch. I didn't managed to see exactly the error that you produced (no tests run), but I did find a variant of #3581: when I use -Dtest, then the test method will run twice - once under the fragment `org.osgi.test.junit5-test`, and once under its host bundle `org.osgi.test.junit5`.

I noticed that the branch had changed a bit from your OP, as there was no `test1()` method in `BundleContextExtensionExampleTest` (only `test2()`). Or perhaps it hadn't changed, and this explains why your command line ran no tests?

Another thing that I'm not sure how it will work is if you try and run a test method that takes parameters - did the `test1()` method that you were running against have any parameters?

**Comment by kriegfrj** [comment]
Ok, it was staring me right in the face - the `test1()` method was there and my eyes scanned past it because it had a parameter. I'm 99% certain that the fact it has a parameter is the cause.

Looking at a fix now.

**Comment by kriegfrj** [comment]
Ok, #3588 fixes the first part of this problem. There were a couple of different issues:

1. If you supply a method selector to any of the sub-engines and it fails to match, then that sub-engine would throw an exception. We're now catching that and hiding it (it might be better if the exception is reported as a test error for the test descriptor for the corresponding engine).
2. You need to specify the parameters for a parameterized test method. The syntax is simply a bracketed, comma-separated list of fqns of the parameter types, ie in your case: `org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1(org.osgi.framework.BundleContext)`
3. Because comma is used as the separator for individual tests in `Test-Cases` and `tester.names`, you need to quote your test method specification if you have more than one parameter.

The above may work for your case I think @rotty3000 , because your parameter is a framework class. However, the more generic case will not work, because the code in `org.junit.platform.engine.discovery` is not OSGi-aware and will try to load the parameter classes from the executing context instead of from the context of the bundle containing the test. To fix this, the `BundleSelectorResolver` will need to resolve the text-based `MethodSelector`s itself and build a new one using the `selectMethod(Class<?>, Method)` variant (similar to what it does for `ClassSelector`) .

**Comment by bjhargrave** [comment]
>     2\. You need to specify the parameters for a parameterized test method. The syntax is simply a bracketed, comma-separated list of fqns of the parameter types, ie in your case: `org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1(org.osgi.framework.BundleContext)`

Generally the command line tools, gradle and maven, do not require you to supply the full method signature. We should find the `test1` method for the user when they do not supply the parameter. If the method name is overloaded, we can pick one or fail.



>     3\. Because comma is used as the separator for individual tests in `Test-Cases` and `tester.names`, you need to quote your test method specification if you have more than one parameter.

For gradle, you just specify `--tests` multiple times, once for each test. A List<String> is assembled by ProjectTester and assembled into `tester.names` by the impls

https://github.com/bndtools/bnd/blob/e5cd53183736ec36254723fcfbaa13e451e2cef2/biz.aQute.junit/src/aQute/junit/plugin/ProjectTesterImpl.java#L57-L61

https://github.com/bndtools/bnd/blob/e5cd53183736ec36254723fcfbaa13e451e2cef2/biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/plugin/ProjectTesterImpl.java#L68-L72

So this code will also need to quote each element as it is assembled into the `tester.names` property.

**Comment by kriegfrj** [comment]
> > ```
> > 2\. You need to specify the parameters for a parameterized test method. The syntax is simply a bracketed, comma-separated list of fqns of the parameter types, ie in your case: `org.osgi.test.junit5.context.BundleContextExtensionExampleTest:test1(org.osgi.framework.BundleContext)`
> > ```
> 
> Generally the command line tools, gradle and maven, do not require you to supply the full method signature. We should find the `test1` method for the user when they do not supply the parameter. If the method name is overloaded, we can pick one or fail.
> 

Is the right place to do this in the tester (which is trying to follow JUnit Platform syntax), or in the gradle/maven plugins?

**Comment by bjhargrave** [comment]
> Is the right place to do this in the tester (which is trying to follow JUnit Platform syntax), or in the gradle/maven plugins?

In the tester. The gradle and maven plugins merely pass the string onto the tester via a `List<String>` argument to `Project.test`. Each tester will need to map that string onto a specific test class and possibly test method.

We don't want to duplicate this mapping in the gradle and maven plugins and they should also be ignorant of the underlying tester engine.

**Comment by kriegfrj** [comment]
Hmm, looks like I didn't enter the comment properly in the commit and this didn't auto-close.

Should have been fixed by #3588
